Rolling node upgrade

mbforbes opened the pull request:
Fixes #6088

After doing this and waiting a bit, `kubectl get nodes` reports nodes as `Unknown` and then back to `Ready`. Also, `kubectl get pods` reports the pods on the minions as `Pending` and then eventually `Running`.

I have some confidence this might actually be the true state of the world, because I screwed up the kube/kube-proxy tokens in an earlier implementation, and neither of the above were true.

BTW, the validate cluster doesn't actually do what's intended right now, I think, because it returned successfully far before my cluster was working again.

Comments / feedback welcome. And yes, definitely want to bake this into a(n e2e) test ASAP.

mbforbes committed:
Rolling node upgrade

mbforbes commented:
+cc @roberthbailey

roberthbailey commented:
For validate-cluster, did you get pull #7932? I noticed this morning that validate-cluster was broken and @fabioy fixed it this afternoon.

roberthbailey commented on cluster/gce/upgrade.sh:
```diff
@@ -102,6 +100,10 @@ function wait-for-master() {
   detect-project
+  # TODO(mbforbes): Do we still need get-password?
+  get-password
+  # TODO(mbforbes): Do we need get-bearer-token here?
```
I don't see why you'd need the kubectl (client) bearer token to upgrade nodes.

mbforbes replied:
`KUBE_BEARER_TOKEN` is required by `build-kube-env`:
https://github.com/GoogleCloudPlatform/kubernetes/blob/9b2a0df811db637442daf93bcffefacaedea6f2a/cluster/gce/debian/helper.sh#L46

However, this comment in `configure-vm.sh`:
https://github.com/GoogleCloudPlatform/kubernetes/blob/9b2a0df811db637442daf93bcffefacaedea6f2a/cluster/gce/configure-vm.sh#L256

says that we don't need these data (`KUBE_PASSWORD`, `KUBE_USER`, `KUBE_BEARER_TOKEN`, `KUBELET_TOKEN`, `KUBE_PROXY_TOKEN`) except on init, and specifically not on an upgrade. However, it might be assuming that an upgrade happens in-place and retains the data that happens on cluster init; with the upgrade process introduced in this PR, that isn't true as we blow away the node.

The call to the function discussed in the previous paragraph, `create-salt-auth`, happens for both master and nodes:
https://github.com/GoogleCloudPlatform/kubernetes/blob/9b2a0df811db637442daf93bcffefacaedea6f2a/cluster/gce/configure-vm.sh#L456

so I don't know whether we can just remove these vars. @zmerlynn should probably chime in here.

roberthbailey replied:
We shouldn't need auth data when updating the master, since the master stores the auth data on a PD which survives rebuilding the VM (and if we lose that data PD we are pretty well screwed anyway since it contains the cluster state). 

For the nodes, you are already scraping the kubelet token and kube proxy token from the existing metadata, so that can be used to fill in those two fields. 

Unless I'm reading this incorrectly, this is within upgrade.sh so it shouldn't affect cluster provisioning to remove these here.

mbforbes replied:
I've moved this to the master-only section since #8009 has been merged.

roberthbailey commented on cluster/gce/upgrade.sh:
```diff
@@ -111,13 +113,57 @@ function prepare-upgrade() {
+  gcloud compute ssh --zone ${ZONE} ${MASTER_NAME} --command \
+    "curl --fail --silent -H 'Metadata-Flavor: Google' \
+      'http://metadata/computeMetadata/v1/instance/attributes/kube-env'" 2>/dev/null \
+    | grep ${1} | cut -d : -f 2 | cut -d \' -f 2
```
does this need a || true at the end so that an error doesn't kill the script?

mbforbes replied:
These vars are required, so it should kill the script if it fails. I can add retries if you want?

roberthbailey replied:
It will kill the script pretty abruptly (and without any sort of error message to the user). Not too worried for the first cut, but can you add a TODO to make this more reliable?

mbforbes replied:
Done.

roberthbailey commented on cluster/gce/upgrade.sh:
```diff
@@ -102,6 +100,10 @@ function wait-for-master() {
 function prepare-upgrade() {
   ensure-temp-dir
   detect-project
+  # TODO(mbforbes): Do we still need get-password?
```
I don't think we need the password for anything. The nodes don't ever get the password, and the master should have it on the data disk even if the boot vm is replaced.

mbforbes replied:
Follow same argument as my lengthy comment above about the bearer token. (And I have the same question—if it's indeed not needed, we need to change some stuff.)

mbforbes replied:
(Done as with above.)

roberthbailey commented on cluster/gce/upgrade.sh:
```diff
@@ -111,13 +113,57 @@ function prepare-upgrade() {
+  gcloud preview rolling-updates start \
+      --group "${NODE_INSTANCE_PREFIX}-group" \
+      --max-num-concurrent-instances 1 \
+      --max-num-failed-instances 1 \
```
Does setting this to 1 mean that we can leave one dead instance behind during the rolling update?

mbforbes replied:
Agh, I read the docs backwards. Thought it meant max num that can fail before the whole thing is failed, not _without_ failing the whole thing. Switched to 0. (I'm going to push without squashing so these comments don't go away.)

roberthbailey commented:
Can you explain what happens during the rolling update? Do the new VMs have different names than the original VMs? Does a new VM get created (and healthchecked in any way) before deleting an existing VM? How does the rolling update command know when to start updating the second VM?

mbforbes commented:
@roberthbailey these answers are from my current understanding of MIG rolling-updates.

**Do the new VMs have different names than the original VMs?**
No; they are recreated under the same name. This we're sure of.

**Does a new VM get created (and healthchecked in any way) before deleting an existing VM?**
We believe that a new VM gets deleted before it is re-created because it is created with the same name, which would otherwise be a naming conflict. 

A rolling-update ensures that the state of the new VM is "running," but it is not healthchecked in any other way, it does not support pluggable healthchecking, and it will not try to repair a VM if it is not "running"—it will simply pause the rolling-update.

**How does the rolling update command know when to start updating the second VM?**
I believe that this can be controlled by setting the `--max-num-concurrent-instances` and `--min-instance-update-time` flags. For example, if you set

- `--max-num-concurrent-instances=1`
- `--min-instance-update-time=0s`

then it would proceed to the second as soon as the first is finished.

What I'm still not clear on is how it determines that the first is finished—whether it waits for the first instance to be "running," or whether it just launches a "create instance" call and moves on, relying a background thread to check that the instances are running.

zmerlynn commented:
LGTM

mbforbes commented:
I think this actually might be actually ready to get in as a draft. I've addressed @roberthbailey's shell comments and pushed my newer version. There are still a handful of cleanup TODOs but the core of what we care about is in. This also fixes master upgrade, which has drifted.

@roberthbailey I've updated my comment response above that answers the questions you posed; let me know if it's adequate. 

@zmerlynn LGTY?

mbforbes commented:
Well, that was fast.
